drivers/gpio/gpio_rpi_pico.c

```diff
@@ -0,0 +1,184 @@
+	}
+
+	return 0;
+}
+
+static int gpio_raspberrypi_port_get_raw(const struct device *dev, uint32_t *value)
+{
+	*value = gpio_get_all();
+	return 0;
+}
+
+static int gpio_raspberrypi_port_set_masked_raw(const struct device *port,
+					uint32_t mask, uint32_t value)
+{
+	gpio_put_masked(mask, value);
+	return 0;
+}
+
+static int gpio_raspberrypi_port_set_bits_raw(const struct device *port,
+					uint32_t pins)
+{
+	gpio_set_mask(pins);
+	return 0;
+}
+
+static int gpio_raspberrypi_port_clear_bits_raw(const struct device *port,
+					uint32_t pins)
+{
+	gpio_clr_mask(pins);
+	return 0;
+}
+
+static int gpio_raspberrypi_port_toggle_bits(const struct device *port,
+					uint32_t pins)
+{
+	gpio_xor_mask(pins);
+	return 0;
+}
+
+static int gpio_raspberrypi_pin_interrupt_configure(const struct device *dev,
+						gpio_pin_t pin,
+						enum gpio_int_mode mode,
+						enum gpio_int_trig trig)
+{
+	struct gpio_raspberrypi_data *data = dev->data;
+	uint32_t events = 0;
+
+	if (mode != GPIO_INT_DISABLE) {
+		if (mode & GPIO_INT_EDGE) {
+			if (trig & GPIO_INT_LOW_0) {
+				events |= GPIO_IRQ_EDGE_FALL;
+			}
+			if (trig & GPIO_INT_HIGH_1) {
+				events |= GPIO_IRQ_EDGE_RISE;
+			}
+		} else {
+			if (trig & GPIO_INT_LOW_0) {
+				events |= GPIO_IRQ_LEVEL_LOW;
+			}
+			if (trig & GPIO_INT_HIGH_1) {
+				events |= GPIO_IRQ_LEVEL_HIGH;
+			}
+		}
+		gpio_set_irq_enabled(pin, events, true);
+	}
+	WRITE_BIT(data->int_enabled_mask, pin, mode != GPIO_INT_DISABLE);
+	return 0;
+}
+
+static int gpio_raspberrypi_manage_callback(const struct device *dev,
+				struct gpio_callback *callback, bool set)
+{
+	struct gpio_raspberrypi_data *data = dev->data;
+
+	return gpio_manage_callback(&data->callbacks, callback, set);
+}
+
+static const struct gpio_driver_api gpio_raspberrypi_driver_api = {
+	.pin_configure = gpio_raspberrypi_configure,
+	.port_get_raw = gpio_raspberrypi_port_get_raw,
+	.port_set_masked_raw = gpio_raspberrypi_port_set_masked_raw,
+	.port_set_bits_raw = gpio_raspberrypi_port_set_bits_raw,
+	.port_clear_bits_raw = gpio_raspberrypi_port_clear_bits_raw,
+	.port_toggle_bits = gpio_raspberrypi_port_toggle_bits,
+	.pin_interrupt_configure = gpio_raspberrypi_pin_interrupt_configure,
+	.manage_callback = gpio_raspberrypi_manage_callback,
+};
+
+static void gpio_raspberrypi_isr(const struct device *dev)
+{
+	struct gpio_raspberrypi_data *data = dev->data;
+	io_irq_ctrl_hw_t *irq_ctrl_base;
+	io_rw_32 *status_reg;
+	uint32_t events;
+	uint32_t pin;
+
+	irq_ctrl_base = &iobank0_hw->proc0_irq_ctrl;
+	for (pin = 0; pin < NUM_BANK0_GPIOS; pin++) {
+		status_reg = &irq_ctrl_base->ints[pin / 8];
+		events = (*status_reg >> 4 * (pin % 8)) & ALL_EVENTS;
+		if (events) {
+			gpio_acknowledge_irq(pin, ALL_EVENTS);
+			gpio_fire_callbacks(&data->callbacks, dev, BIT(pin));
+		}
+	}
+}
+
+static int gpio_raspberrypi_init(const struct device *dev)
+{
+	IRQ_CONNECT(IO_IRQ_BANK0, IRQ_PRIORITY, gpio_raspberrypi_isr,
+			DEVICE_DT_GET(DT_INST(0, raspberrypi_pico_gpio)), 0);
```

Comment:
Something is wrong here. This function implies that the driver is single-instance, but below it's using DT_INST_FOREACH_STATUS_OKAY. What's going on? It looks like this should be moved into the GPIO_RASPBERRYPI_INIT macro if you have a potential multi-instance situation in the future.